executor/analyze.go

```diff
@@ -502,16 +533,450 @@ type AnalyzeFastExec struct {
+		collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos].Value = types.NewBytesDatum(bytes)
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleBatchGetResponse(kvMap map[string][]byte, collectors []*statistics.SampleCollector) (err error) {
+	length := int32(len(kvMap))
+	newCursor := atomic.AddInt32(&e.sampCursor, length)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	samplePos := newCursor - length
+	for sKey, sValue := range kvMap {
+		err = e.updateCollectorSamples(collectors, sValue, kv.Key(sKey), samplePos, hasPKInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		samplePos++
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleScanIter(iter kv.Iterator, collectors []*statistics.SampleCollector) (err error) {
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	rander := rand.New(rand.NewSource(RandSeed + int64(e.rowCount)))
+	for ; iter.Valid() && err == nil; err = iter.Next() {
+		// reservoir sampling
+		e.rowCount++
+		randNum := rander.Int63n(int64(e.rowCount))
+		if randNum > int64(MaxSampleSize) && e.sampCursor == int32(MaxSampleSize) {
+			continue
+		}
+
+		p := rander.Int31n(int32(MaxSampleSize))
+		if e.sampCursor < int32(MaxSampleSize) {
+			p = e.sampCursor
+			e.sampCursor++
+		}
+
+		err = e.updateCollectorSamples(collectors, iter.Value(), iter.Key(), p, hasPKInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return errors.Trace(err)
+}
+
+func (e *AnalyzeFastExec) handleScanTasks(bo *tikv.Backoffer, collectors []*statistics.SampleCollector) error {
+	snapshot, err := e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, t := range e.scanTasks {
+		iter, err := snapshot.Iter(t.StartKey, t.EndKey)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = e.handleScanIter(iter, collectors)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, rid int, err *error, collectors []*statistics.SampleCollector) {
+	defer func() {
+		e.wg.Done()
+		*err = errors.Trace(*err)
+	}()
+	var snapshot kv.Snapshot
+	snapshot, *err = e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	rander := rand.New(rand.NewSource(RandSeed + int64(rid)))
+	if *err != nil {
+		return
+	}
+	for i := rid; i < len(e.sampTasks); i += e.concurrency {
+		task := e.sampTasks[i]
+		if task.SampSize == 0 {
+			continue
+		}
+
+		var tableID, minRowID, maxRowID int64
+		startKey, endKey := task.Location.StartKey, task.Location.EndKey
+		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		if *err != nil {
+			return
+		}
+		_, maxRowID, *err = tablecodec.DecodeRecordKey(endKey)
+		if *err != nil {
+			return
+		}
+
+		keys := make([]kv.Key, 0, task.SampSize)
+		for i := 0; i < int(task.SampSize); i++ {
+			randKey := rander.Int63n(maxRowID-minRowID) + minRowID
```

Comment:
Since rand.Intn(0) will result in panic, do we need to check `maxRowID-minRowID > 0` here?